lib/cosmos/dart/script/dart_reducer.rb

```diff
@@ -0,0 +1,277 @@
+              time_attribute = :yday
+            else
+              Cosmos::Logger.error("Reducer Thread Unexpected Job Type: #{job_type}")
+            end
+
+            rows = []
+            model1.where("reduced_state = 0").order("meta_id ASC, #{model1_time_column} ASC").find_each do |row|
+              rows << row
+              if (rows[-1].send(model1_time_column) - rows[0].send(model1_time_column)) > time_delta or rows[0].send(model1_time_column).send(time_attribute) != rows[-1].send(model1_time_column).send(time_attribute) or rows[0].meta_id != rows[-1].meta_id
+                sample_rows = rows[0..-2]
+                new_row = model2.new
+                new_row.start_time = sample_rows[0].send(model1_time_column)
+                new_row.num_samples = sample_rows.length
+                new_row.meta_id = sample_rows[0].meta_id
+                mappings.each do |mapping|
+                  item_name = "i#{mapping.item_index}"
+                  max_item_name = "i#{mapping.item_index}max"
+                  min_item_name = "i#{mapping.item_index}min"
+                  avg_item_name = "i#{mapping.item_index}avg"
+                  max_value = nil
+                  min_value = nil
+                  avg_value = 0.0
+                  total_samples = 0
+                  max_nan_found = false
+                  min_nan_found = false
+                  avg_nan_found = false
+                  sample_rows.each do |row_to_reduce|
+                    if job_type == :MINUTE
+                      value = row_to_reduce.read_attribute(item_name)
+                      max_sample = value
+                      min_sample = value
+                      avg_sample = value
+                      if value.nil?
+                        Cosmos::Logger.error("#{item_name} is nil in #{row_to_reduce.class}:#{row_to_reduce.id}")
+                        next
+                      end
+                    else
+                      max_sample = row_to_reduce.read_attribute(max_item_name)
+                      min_sample = row_to_reduce.read_attribute(min_item_name)
+                      avg_sample = row_to_reduce.read_attribute(avg_item_name)
+                      if max_sample.nil?
+                        Cosmos::Logger.error("#{max_item_name} is nil in #{row_to_reduce.class}:#{row_to_reduce.id}")
+                        next
+                      end
+                      if min_sample.nil?
+                        Cosmos::Logger.error("#{min_item_name} is nil in #{row_to_reduce.class}:#{row_to_reduce.id}")
+                        next
+                      end
+                      if avg_sample.nil?
+                        Cosmos::Logger.error("#{avg_item_name} is nil in #{row_to_reduce.class}:#{row_to_reduce.id}")
+                        next
+                      end
+                    end
+
+                    unless (max_sample.is_a?(Float) and (avg_sample.nan? or !avg_sample.finite?))
```

Comment:
I don't understand this logic. If the min is a float and the avg is non a number we say min_nan_found. Otherwise we determine if the value should be the new minimum. Shouldn't the check simply be whether the min_sample is nan or not finite. What does the avg_sample have to do with it?

Reply:
Ok I think this is a typo as well as the check for min. I'll fix in my documentation / refactor.